# Классная работа 6

## Раздел 1

Пункты 1-2

In [18]:
import pandas as pd
import sklearn.model_selection as skms
import sklearn.linear_model as sklm
import sklearn.metrics as skmt
filename = 'data/pima-indians-diabetes.data.csv'
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
data = pd.read_csv(filename, names = names)
data.head()

preg  plas  pres  skin  test  mass   pedi  age  class
0     6   148    72    35     0  33.6  0.627   50      1
1     1    85    66    29     0  26.6  0.351   31      0
2     8   183    64     0     0  23.3  0.672   32      1
3     1    89    66    23    94  28.1  0.167   21      0
4     0   137    40    35   168  43.1  2.288   33      1

Пункты 3-4

In [19]:
array = data.values
x = array[:, :-1]
y = array[:, -1]

X - признаки, предположительно влияющие на статус диабета

Y - статус диабета

Пункты 5-8

In [20]:
test_size = 0.33
random_state = 7
x_train, x_test, y_train, y_test = skms.train_test_split(x, y, test_size=test_size, random_state=random_state)
model1 = sklm.LogisticRegression(solver='liblinear')
model1.fit(x_train, y_train)
result_sc = model1.score(x_test, y_test)
print(result_sc)

0.7559055118110236


Пункты 9-11

In [21]:
n_splits = 10
random_state = 7
kfold = skms.KFold(n_splits= n_splits, random_state = random_state, shuffle=True)
result_cv = skms.cross_val_score(model1, x, y, cv=kfold)
print(result_cv.mean(), result_cv.std())

0.7708646616541353 0.05090500786917546


Пункты 12-14

In [22]:
loocv = skms.LeaveOneOut()
result_cv2 = skms.cross_val_score(model1, x, y, cv=loocv)
print(result_cv2.mean(), result_cv2.std())

0.7682291666666666 0.42196340380334596


Пункты 15-17

In [23]:
kfold3 = skms.ShuffleSplit(n_splits = 10, test_size = 0.33, random_state = 7)
result_cv3 = skms.cross_val_score(model1, x, y, cv=kfold)
print(result_cv3.mean(), result_cv3.std())

0.7708646616541353 0.05090500786917546


## Раздел 2

Пункты 18-20

In [24]:
kfold4 = skms.KFold(n_splits = 10, random_state = 7, shuffle=True)
result_cv4 = skms.cross_val_score(model1, x, y, cv=kfold4, scoring='accuracy')
print(result_cv4.mean(), result_cv4.std())

0.7708646616541353 0.05090500786917546


Пункты 21-22

In [25]:
result_cv5 = skms.cross_val_score(model1, x, y, cv=kfold4, scoring='neg_log_loss')
print(-result_cv5.mean(), result_cv5.std())

0.49356405689112515 0.04206587818993993


Пункты 23-24

In [26]:
result_cv6 = skms.cross_val_score(model1, x, y, cv=kfold4, scoring='roc_auc')
print(result_cv6.mean(), result_cv6.std())

0.8257627921677674 0.05005520303752424


Пункт 25-27

In [27]:
predicted = model1.predict(x_test)
confusion_m = skmt.confusion_matrix(y_test, predicted)
print(confusion_m)

[[141  21]
 [ 41  51]]


Результативный признак class имеет два значения: 1 и 0.

[141 21] - корректно определенные значения для класса 1 и 0 соответственно

[41 51] - некорректно определенные значения для класса 1 и 0 соответственно

Пункты 28-29

In [28]:
report = skmt.classification_report(y_test, predicted)
print(report)

              precision    recall  f1-score   support

         0.0       0.77      0.87      0.82       162
         1.0       0.71      0.55      0.62        92

    accuracy                           0.76       254
   macro avg       0.74      0.71      0.72       254
weighted avg       0.75      0.76      0.75       254


В результате получем числовые значения следующих параметров:

precision - отношение true positive предположений и общего числа positive предположений

recall - отношение всех корректных positive предположений и общего числа реальных positive

f1-score - сбалансированное среднее precision и recall

support - кол-во появлений класса в датасете

accuracy - общая точность предположений


## Раздел 3

Пункт 30

In [29]:
filename = "data/housing.csv"
names = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO','B', 'LSTAT', 'MEDV']
data2 = pd.read_csv(filename, delim_whitespace=True, names=names)
data2.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

Пункты 31-32

In [30]:
array2 = data2.values
x2 = array[:, :-1]
y2 = array[:, -1]

X - признаки, предположительно влияющие на среднюю стоимость недвижимости в Бостоне

Y - стоимость недвижимости в Бостоне

Пункты 33-35

In [31]:
model2 = sklm.LinearRegression()
kfold2 = skms.KFold(n_splits = 10, random_state = 7, shuffle=True)
result2_cv = skms.cross_val_score(model2, x2, y2, cv=kfold2, scoring='neg_mean_absolute_error')
print(-result2_cv.mean(), result2_cv.std())

0.3368404421024537 0.019780732404569353


MAE измеряет разницу между предположенными и реальными значениями. Чем ниже MAE, тем лучше. В данном примере оно составляет ~0.34 тыс. долл., т.е. 340 долл.

Второе значение - разброс. Чем ниже, тем выше постоянство модели.

Можно заключить, что кросс-валидация подтвердила эффективность работы модели: значение MAE и его разброс довольно низки.

Пункты 36-37

In [32]:
result2_cv2 = skms.cross_val_score(model2, x2, y2, cv=kfold2, scoring='neg_mean_squared_error')
print(-result2_cv2.mean(), result2_cv2.std())

0.16320738575907065 0.020251428034588133


MSE так же рассчитывает разницу между предположенными и реальными значениями, но расчет метрики проводится по-другому. Значение будет представлено в у. ед. в квадрате, для нас - тыс. долл. в квадрате.

Пункты 38-39

In [33]:
result2_cv3 = skms.cross_val_score(model2, x2, y2, cv=kfold2, scoring='r2')
print(result2_cv3.mean(), result2_cv3.std())

0.26432881386388385 0.10236184260132886


R^2 показывает, насколько регрессионная модель подходит под датасет.

1 - полностью подходит
0 - не подходит

В отличие от предыдуших кросс-валидаций данная показала более пессемистичный результат.